homeassistant/components/lifx/light.py

```diff
@@ -376,72 +393,122 @@ async def start_effect(self, entities, service, **kwargs):
         elif service == SERVICE_EFFECT_STOP:
             await self.effects_conductor.stop(bulbs)
 
+    def clear_inflight_discovery(self, inflight: InFlightDiscovery) -> None:
+        """Clear in-flight discovery."""
+
+        if inflight.device.mac_addr in self.discoveries_inflight:
+            self.discoveries_inflight.pop(inflight.device.mac_addr)
+
     @callback
-    def register(self, bulb):
+    def register(self, bulb: Light) -> None:
         """Allow a single in-flight discovery per bulb."""
+        if bulb.mac_addr in self.switch_devices:
+            _LOGGER.debug(
+                "Skipping discovered LIFX Switch at %s (%s)",
+                bulb.ip_addr,
+                bulb.mac_addr,
+            )
+            return
+
         if bulb.mac_addr not in self.discoveries_inflight:
-            self.discoveries_inflight[bulb.mac_addr] = bulb.ip_addr
-            _LOGGER.debug("Discovered %s (%s)", bulb.ip_addr, bulb.mac_addr)
-            self.hass.async_create_task(self.register_bulb(bulb))
+            inflight = InFlightDiscovery(bulb, asyncio.Lock())
+            self.discoveries_inflight[bulb.mac_addr] = inflight
+            _LOGGER.debug(
+                "Discovery response received from %s (%s)",
+                bulb.ip_addr,
+                bulb.mac_addr,
+            )
         else:
-            _LOGGER.warning("Duplicate LIFX discovery response ignored")
+            _LOGGER.debug(
+                "Duplicate discovery response received from %s (%s)",
+                bulb.ip_addr,
+                bulb.mac_addr,
+            )
+
+        self.hass.async_create_task(
+            self._async_handle_discovery(self.discoveries_inflight[bulb.mac_addr])
+        )
 
-    async def register_bulb(self, bulb):
+    async def _async_handle_discovery(self, inflight: InFlightDiscovery) -> None:
         """Handle LIFX bulb registration lifecycle."""
-        if bulb.mac_addr in self.entities:
-            entity = self.entities[bulb.mac_addr]
-            entity.registered = True
-            _LOGGER.debug("Reconnected to %s", entity.who)
-            await entity.update_hass()
-        else:
-            _LOGGER.debug("Connecting to %s (%s)", bulb.ip_addr, bulb.mac_addr)
 
-            # Read initial state
+        # only allow a single discovery process per discovered device
+        async with inflight.lock:
+
```

Comment:
We should check that the previous run didn't create the entity after we acquire the lock.

Reply:
I never delete the entity, so I don't need to recreate it. It always exists, I just change it's availability status.